pkg/operator/endpoint_status_builder.go

```diff
@@ -68,24 +68,132 @@ func (b *scrapeEndpointBuilder) add(target *prometheusv1.TargetsResult) error {
 	return nil
 }
 
+func setNamespacedObjectByScrapeJobKey(o monitoringv1.PodMonitoringCRD, split []string, full string) (monitoringv1.PodMonitoringCRD, error) {
+	if len(split) != 3 {
+		return nil, fmt.Errorf("invalid %s scrape key format %q", split[0], full)
+	}
+
+	o.SetNamespace(split[1])
+	o.SetName(split[2])
+	return o, nil
+}
+
+func setClusterScopedObjectByScrapeJobKey(o monitoringv1.PodMonitoringCRD, split []string, full string) (monitoringv1.PodMonitoringCRD, error) {
+	if len(split) != 2 {
+		return nil, fmt.Errorf("invalid %s scrape key format %q", split[0], full)
+	}
+
+	o.SetName(split[1])
+	return o, nil
+}
+
+// getObjectByScrapeJobKey converts the key to a CRD. See monitoringv1.PodMonitoringCRD.GetKey().
+func getObjectByScrapeJobKey(key string) (monitoringv1.PodMonitoringCRD, error) {
+	split := strings.Split(key, "/")
+	// Generally:
+	// - "kind" for scrape pools without a respective CRD.
+	// - "kind/name" for cluster-scoped resources.
+	// - "kind/namespace/name" for namespaced resources.
+	switch split[0] {
+	case "kubelet":
+		if len(split) != 1 {
+			return nil, fmt.Errorf("invalid kubelet scrape key format %q", key)
+		}
+		return nil, nil
+	case "PodMonitoring":
+		return setNamespacedObjectByScrapeJobKey(&monitoringv1.PodMonitoring{}, split, key)
+	case "ClusterPodMonitoring":
+		return setClusterScopedObjectByScrapeJobKey(&monitoringv1.ClusterPodMonitoring{}, split, key)
+	case "NodeMonitoring":
+		if _, err := setClusterScopedObjectByScrapeJobKey(&monitoringv1.ClusterPodMonitoring{}, split, key); err != nil {
+			return nil, err
+		}
+		return nil, nil
+	default:
+		return nil, fmt.Errorf("unknown scrape kind %q", split[0])
+	}
+}
+
+// scrapePool is the parsed Prometheus scrape pool, which is assigned to the job name from our
+// configurations. For example, for PodMonitoring this is `PodMonitoring/namespace/name/port`. The
+// key is what identifies the resource (`PodMonitoring/namespace/name`) and the group indicates a
+// small subset of that resource (`port`).
+type scrapePool struct {
+	key   string
+	group string
+}
+
+// slicesJoinLength is strings.Join but only returns the length of the resulting string.
+func stringsJoinLength(elems []string, sep string) int {
+	if len(elems) == 0 {
+		return 0
+	}
+	j := len(elems) - 1
+	for _, elem := range elems {
+		j += len(elem)
+	}
+	return j
+}
+
+func parseScrapePool(pool string) (scrapePool, error) {
+	split := strings.Split(pool, "/")
+	switch split[0] {
+	case "kubelet":
+		if len(split) != 2 {
+			return scrapePool{}, fmt.Errorf("invalid kubelet scrape pool format %q", pool)
+		}
+		return scrapePool{
+			key:   split[0],
+			group: split[1],
+		}, nil
+	case "PodMonitoring":
+		if len(split) != 4 {
+			return scrapePool{}, fmt.Errorf("invalid PodMonitoring scrape pool format %q", pool)
+		}
+		index := stringsJoinLength(split[:3], "/")
```

Comment:
what do you think about using `index := strings.LastIndex(scrapePool, "/")` ?

Reply:
Great catch! Mostly because this won't work in the future.

For Probe, this will not be the last index. In Probe, the way it currently works is you specify multiple URLs per protocol. For example, you might probe `google.com` and `example.com` using HTTP and `monitoring.googleapis.com:443` for gRPC in the same Probe object. The way we create this group is `protocol/index`, e.g. for `example.com`, `http_2xx/1` (with 1 because it is an array and that is its position `["google.com", "example.com"]`). A workaround might be to avoid using a slash here but I feel that using the same delimiter all throughout is the cleanest.